doc/devel/development_workflow.rst

```diff
@@ -358,3 +287,131 @@ and the history looks now like this::
+
+See `rebase without tears`_ for more detail.
+
+.. _rebase without tears: https://matthew-brett.github.io/pydagogue/rebase_without_tears.html
+
+To do a rebase on ``upstream/main``::
+
+    # Fetch changes from upstream/main
+    git fetch upstream
+    # go to the feature branch
+    git checkout cool-feature
+    # make a backup in case you mess up
+    git branch tmp cool-feature
+    # rebase cool-feature onto main
+    git rebase --onto upstream/main upstream/main cool-feature
+
+In this situation, where you are already on branch ``cool-feature``, the last
+command can be written more succinctly as::
+
+    git rebase upstream/main
+
+When all looks good you can delete your backup branch::
+
+   git branch -D tmp
+
+If it doesn't look good you may need to have a look at
+:ref:`recovering-from-mess-up`.
+
+If you have made changes to files that have also changed in ``main``, this may
+generate merge conflicts that you need to resolve - see the `git rebase`_ man
+page for some instructions at the end of the "Description" section. There is
+some related help on merging in the git user manual - see `resolving a merge`_.
+
+.. _git rebase: https://git-scm.com/docs/git-rebase
+.. _resolving a merge: https://schacon.github.io/git/user-manual.html#resolving-a-merge
+
+
+If you have not yet pushed this branch to github, you can carry on as normal,
+however if you *have* already pushed this commit see :ref:`force-push` for how
+to replace your already published commits with the new ones.
+
+
+.. _force-push:
+
+
+Pushing, with force
+-------------------
+
+If you have in some way re-written already pushed history (e.g. via
+:ref:`rewriting-commit-history` or :ref:`rebase-on-main`), then when you try to
+push the new commits to GitHub it will fail with an error that looks like ::
+
+   $ git push
+   Pushing to github.com:YOURFORK/matplotlib.git
+   To github.com:YOURFORK/matplotlib.git
+    ! [rejected]              cool_feature -> cool_feature (non-fast-forward)
+   error: failed to push some refs to 'github.com:tacaswell/matplotlib.git'
+   hint: Updates were rejected because the tip of your current branch is behind
+   hint: its remote counterpart. Integrate the remote changes (e.g.
+   hint: 'git pull ...') before pushing again.
+   hint: See the 'Note about fast-forwards' in 'git push --help' for details.
+
+What is going on is that by default ``git push`` tries to protect you from
+accidentally make commits inaccessible by leaving no reference to them so if a
+push *would* effectively discard commits on the remote, ``git`` rejects the
+push.  When this happens GitHub adds the helpful suggestion to pull the remote
+changes and then try pushing again.  In some cases, such as if you and a
+colleague are both committing and pushing to the same branch, this is the
+correct course of action (or rebase your local branch but it is a matter of
+taste).  Although both of things are good ideas, taken together they are not
+what we want to do.
```

Comment:
```suggestion
What is going on is that by default ``git push`` tries to protect you from
accidentally discarding commits on the remote by rejecting the
push.  When this happens, GitHub adds the helpful suggestion to pull the remote
changes and then try pushing again.  In some cases, such as if you and a
colleague are both committing and pushing to the same branch, this is the
correct course of action (or rebase your local branch but it is a matter of
taste). 
```

Reply:
They are the same, but I was trying not to use the same slightly long phrase twice in the same sentence so I used two different overly long phrases in the same sentence.

Reply:
I think for technical writing reusing the same phrase for a single concept is usually good.    I once wrote a paper where I called the same phenomena three different names - it did not go well in review ;-)